inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp

```diff
@@ -43,37 +44,77 @@ void MKLDNNMemoryOutputNode::initSupportedPrimitiveDescriptors() {
     supportedPrimitiveDescriptors.emplace_back(config, impl_desc_type::unknown, memory::format::any);
 }
 
-const MKLDNNEdgePtr MKLDNNMemoryOutputNode::getChildEdgeAt(size_t idx) const {
-    if (inputNode != nullptr) {
-        return inputNode->getChildEdgeAt(idx);
-    }
-    return MKLDNNNode::getChildEdgeAt(idx);
-}
-
 void MKLDNNMemoryOutputNode::execute(mkldnn::stream strm)  {
     auto& srcMemory = getParentEdgeAt(0)->getMemory();
 
-    const float *src_ptr = reinterpret_cast<const float*>(srcMemory.GetData()) +
-            srcMemory.GetDescriptor().data.layout_desc.blocking.offset_padding;
-    float *dst_ptr = reinterpret_cast<float*>(getChildEdgeAt(0)->getMemory().GetData()) +
-            getChildEdgeAt(0)->getMemory().GetDescriptor().data.layout_desc.blocking.offset_padding;
-
-    // TODO: this can be eliminated by completely removing MKLDNN memory output NODE, to fuse it with output of prev layer
-    memcpy(dst_ptr, src_ptr, srcMemory.GetSize());
+    auto inputMemoryNode = dynamic_cast<MKLDNNMemoryInputNode*>(inputNode);
+    IE_ASSERT(inputMemoryNode != nullptr);
+    inputMemoryNode->storeState(srcMemory);
 }
 
 #if defined (COMPILED_CPU_MKLDNN_INPUT_NODE)
 MKLDNNMemoryInputNode::MKLDNNMemoryInputNode(const InferenceEngine::CNNLayerPtr& layer, const mkldnn::engine& eng, MKLDNNWeightsSharing::Ptr &cache)
-        : MKLDNNInputNode(layer, eng, cache), MKLDNNMemoryNode(layer) {
+        : MKLDNNInputNode(layer, eng, cache), MKLDNNMemoryNode(layer), dataStore(new MKLDNNMemory{eng}) {
     if (created()) {
         holder = MKLDNNMemoryNodeVirtualEdge::registerInput(this);
+
+        auto dims = layer->outData[0]->getDims();
+        auto prc  = layer->outData[0]->getPrecision();
+        auto tdesc = TensorDesc(prc, dims, TensorDesc::getLayoutByDims(dims));
+        dataStore->Create(MKLDNNMemoryDesc {tdesc});
+
+        // default memory state is zero filled
+        dataStore->FillZero();
     }
 }
 
+/**
+ * Copy data from one tensor into other.
+ * As is. Assume that data is dense tensor with same layout.
+ * @param dst_mem destination memory object
+ * @param src_mem source memory object
+ */
+inline
+static void simple_copy(MKLDNNMemory& dst_mem, const MKLDNNMemory& src_mem) {
```

Comment:
Could you please minimize places were mixed code styles are used?

Reply:
Improved